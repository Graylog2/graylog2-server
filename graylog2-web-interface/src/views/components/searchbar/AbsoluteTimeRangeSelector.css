:local(.selectorContainer) {
    display: flex;
}

:local(.separator) {
    margin: 0;
    line-height: 34px;
    font-size: 18px;
    padding-left: 15px;
    padding-right: 15px;
}

:local(.inputWidth) {
    width: 200px;
}
