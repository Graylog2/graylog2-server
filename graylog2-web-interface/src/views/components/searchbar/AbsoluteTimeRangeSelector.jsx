import React from 'react';
import PropTypes from 'prop-types';
import Immutable from 'immutable';

import { Button } from 'components/graylog';
import Input from 'components/bootstrap/Input';
import { DatePicker, Icon } from 'components/common';
import DateTime from 'logic/datetimes/DateTime';

import styles from './AbsoluteTimeRangeSelector.css';

const _formattedDateStringInUserTZ = (dateString) => {
  return DateTime.parseFromString(dateString);
};

const _setDateTimeToNow = (field, onChange) => {
  const value = new DateTime().toISOString();
  onChange(field, value);
};

const _isValidDateString = (dateString) => {
  try {
    if (dateString !== undefined) {
      DateTime.parseFromString(dateString);
    }
    return true;
  } catch (e) {
    return false;
  }
};

const _onDateSelected = (date, key, onChange) => {
  const midnightDate = date.setHours(0);
  const newValue = DateTime.ignoreTZ(midnightDate).toISOString();
  onChange(key, newValue);
};

const _extractStateFromProps = (props) => {
  const { from, to } = props.value.map(value => _formattedDateStringInUserTZ(value).toString()).toObject();
  return {
    from,
    to,
  };
};

export default class AbsoluteTimeRangeSelector extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = _extractStateFromProps(props);
  }

  componentWillReceiveProps(nextProps) {
    const { value } = this.props;

    if (!Immutable.is(nextProps.value, value)) {
      this.setState(_extractStateFromProps(nextProps));
    }
  }

  onBlur = (key, onChange) => {
    // eslint-disable-next-line react/destructuring-assignment
    const value = this.state[key];

    if (_isValidDateString(value)) {
      onChange(key, _formattedDateStringInUserTZ(value).toISOString());
    }
  };

  onChange = (key, value) => {
    this.setState({ [key]: value });
  };

  render() {
    const { from, to } = this.state;
    const { disabled, onChange } = this.props;

    return (
      <div className={`timerange-selector absolute ${styles.selectorContainer}`}>
        <input type="hidden" name="from" />
        <div className={styles.inputWidth}>
          <DatePicker id="searchFromDatePicker"
                      disabled={disabled}
                      title="Search start date"
                      date={from}
                      onChange={date => _onDateSelected(date, 'from', onChange)}>
            <Input id="fromDateInput"
                   type="text"
                   disabled={disabled}
                   className="absolute"
                   value={from}
                   onBlur={() => this.onBlur('from', onChange)}
                   onChange={event => this.onChange('from', event.target.value)}
                   placeholder={DateTime.Formats.DATETIME}
                   buttonAfter={(
                     <Button disabled={disabled} onClick={() => _setDateTimeToNow('from', onChange)}>
                       <Icon name="magic" />
                     </Button>
                   )}
                   bsStyle={_isValidDateString(from) ? null : 'error'}
                   required />
          </DatePicker>
        </div>

        <p className={`text-center ${styles.separator}`}>
          <Icon name="long-arrow-right" />
        </p>
        <input type="hidden" name="to" />
        <div className={styles.inputWidth}>
          <DatePicker id="searchToDatePicker"
                      disabled={disabled}
                      title="Search end date"
                      date={to}
                      onChange={date => _onDateSelected(date, 'to', onChange)}>
            <Input id="toDateInput"
                   type="text"
                   disabled={disabled}
                   className="absolute"
                   value={to}
                   onBlur={() => this.onBlur('to', onChange)}
                   onChange={event => this.onChange('to', event.target.value)}
                   placeholder={DateTime.Formats.DATETIME}
                   buttonAfter={(
                     <Button disabled={disabled} onClick={() => _setDateTimeToNow('to', onChange)}>
                       <Icon name="magic" />
                     </Button>
                   )}
                   bsStyle={_isValidDateString(to) ? null : 'error'}
                   required />
          </DatePicker>
        </div>
      </div>
    );
  }
}

AbsoluteTimeRangeSelector.propTypes = {
  disabled: PropTypes.bool,
  value: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
};

AbsoluteTimeRangeSelector.defaultProps = {
  disabled: false,
};
