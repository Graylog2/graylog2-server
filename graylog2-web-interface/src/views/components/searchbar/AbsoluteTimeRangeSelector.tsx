/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'formik';
import styled, { css } from 'styled-components';

import DateTime from 'logic/datetimes/DateTime';
import { Icon } from 'components/common';

import TimerangeSelector from './TimerangeSelector';
import DateInputWithPicker from './DateInputWithPicker';

type Props = {
  disabled: boolean,
};

const StyledTimerangeSelector = styled(TimerangeSelector)`
  display: flex;
`;

const InputWrap = styled.div`
  width: 200px;
`;

const Separator = styled.p(({ theme }) => css`
  margin: 0;
  line-height: 34px;
  font-size: ${theme.fonts.size.large};
  padding-left: 15px;
  padding-right: 15px;
`);

const _isValidDateString = (dateString: string) => {
  if (dateString === undefined) {
    return undefined;
  }

  return DateTime.isValidDateString(dateString)
    ? undefined
    : 'Format must be: YYYY-MM-DD [HH:mm:ss[.SSS]]';
};

const AbsoluteTimeRangeSelector = ({ disabled }: Props) => {
  return (
    <StyledTimerangeSelector className="absolute">
      <Field name="timerange.from" validate={_isValidDateString}>
        {({ field: { value, onChange, onBlur, name }, meta: { error } }) => (
          <InputWrap>
            <DateInputWithPicker disabled={disabled}
                                 onChange={onChange}
                                 onBlur={onBlur}
                                 value={value}
                                 name={name}
                                 title="Search start date"
                                 error={error} />
          </InputWrap>
        )}
      </Field>

      <Separator className="text-center">
        <Icon name="long-arrow-alt-right" />
      </Separator>

      <Field name="timerange.to" validate={_isValidDateString}>
        {({ field: { value, onChange, onBlur, name }, meta: { error } }) => (
          <InputWrap>
            <DateInputWithPicker disabled={disabled}
                                 onChange={onChange}
                                 onBlur={onBlur}
                                 value={value}
                                 name={name}
                                 title="Search end date"
                                 error={error} />
          </InputWrap>
        )}
      </Field>
    </StyledTimerangeSelector>
  );
};

AbsoluteTimeRangeSelector.propTypes = {
  disabled: PropTypes.bool,
};

AbsoluteTimeRangeSelector.defaultProps = {
  disabled: false,
};

export default AbsoluteTimeRangeSelector;
