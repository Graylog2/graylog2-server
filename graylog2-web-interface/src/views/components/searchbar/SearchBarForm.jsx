// @flow strict
import * as React from 'react';
import PropTypes from 'prop-types';
import { useCallback } from 'react';
import { Form, Formik } from 'formik';
import { isFunction } from 'lodash';
import type { FormikProps } from 'formik/@flow-typed';

import DateTime from 'logic/datetimes/DateTime';
import type { TimeRange } from 'views/logic/queries/Query';
import { onInitializingTimerange, onSubmittingTimerange } from 'views/components/TimerangeForForm';

type Values = {
  timerange: TimeRange,
  streams: Array<string>,
  queryString: string,
};

type Props = {
  initialValues: Values,
  onSubmit: (Values) => void | Promise<any>,
  children: ((props: FormikProps<Values>) => React$Node) | React$Node,
};

const validate = (values) => {
  const errors = {};

  if (values.timerange.type === 'absolute' && DateTime.isValidDateString(values.timerange.from) && values.timerange.from > values.timerange.to) {
    errors.timerange = {
      from: 'Start date must be before end date',
    };
  }

  return errors;
};

const SearchBarForm = ({ initialValues, onSubmit, children }: Props) => {
  const _onSubmit = useCallback(({ timerange, streams, queryString }) => {
    const newTimerange = onSubmittingTimerange(timerange);

    return onSubmit({
      timerange: newTimerange,
      streams,
      queryString,
    });
  }, [onSubmit]);
  const { timerange, streams, queryString } = initialValues;
  const _initialValues = {
    queryString,
    streams,
    timerange: onInitializingTimerange(timerange),
  };

  return (
    <Formik initialValues={_initialValues}
            enableReinitialize
            onSubmit={_onSubmit}
            validate={validate}>
      {(...args) => (
        <Form>
          {isFunction(children) ? children(...args) : children}
        </Form>
      )}
    </Formik>
  );
};

SearchBarForm.propTypes = {
  initialValues: PropTypes.shape({
    timerange: PropTypes.object.isRequired,
    queryString: PropTypes.string.isRequired,
    streams: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
  onSubmit: PropTypes.func.isRequired,
};

export default SearchBarForm;
