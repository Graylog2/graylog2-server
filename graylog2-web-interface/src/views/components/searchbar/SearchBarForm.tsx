/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import { useCallback, useContext, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import type { FormikProps } from 'formik';
import { Form, Formik } from 'formik';
import isFunction from 'lodash/isFunction';

import { onInitializingTimerange, onSubmittingTimerange } from 'views/components/TimerangeForForm';
import type { SearchBarFormValues } from 'views/Constants';
import FormWarningsContext from 'contexts/FormWarningsContext';
import type { QueryValidationState } from 'views/components/searchbar/queryvalidation/types';
import validate from 'views/components/searchbar/validate';
import usePluginEntities from 'hooks/usePluginEntities';
import useUserDateTime from 'hooks/useUserDateTime';
import useHandlerContext from 'views/components/useHandlerContext';

type FormRenderer = (props: FormikProps<SearchBarFormValues>) => React.ReactNode;
type Props = {
  children: FormRenderer | React.ReactNode,
  initialValues: SearchBarFormValues,
  limitDuration: number,
  onSubmit: (values: SearchBarFormValues) => Promise<any>,
  validateOnMount?: boolean,
  formRef?: React.Ref<FormikProps<SearchBarFormValues>>,
  validateQueryString: (values: SearchBarFormValues) => Promise<QueryValidationState>,
}

const _isFunction = (children: Props['children']): children is FormRenderer => isFunction(children);

export const normalizeSearchBarFormValues = ({ timerange, ...rest }: SearchBarFormValues, userTimezone: string) => ({ timerange: onSubmittingTimerange(timerange, userTimezone), ...rest });

const SearchBarForm = ({ initialValues, limitDuration, onSubmit, children, validateOnMount, formRef, validateQueryString }: Props) => {
  const [enableReinitialize, setEnableReinitialize] = useState(true);
  const { formatTime, userTimezone } = useUserDateTime();
  const pluggableSearchBarControls = usePluginEntities('views.components.searchBar');
  const { setFieldWarning } = useContext(FormWarningsContext);
  const _onSubmit = useCallback((values: SearchBarFormValues) => {
    setEnableReinitialize(false);

    return onSubmit(normalizeSearchBarFormValues(values, userTimezone)).finally(() => setEnableReinitialize(true));
  }, [onSubmit, userTimezone]);
  const _initialValues = useMemo(() => {
    const { timerange, ...rest } = initialValues;

    return ({
      ...rest,
      timerange: onInitializingTimerange(timerange, formatTime),
    });
  }, [formatTime, initialValues]);

  const handlerContext = useHandlerContext();
  const _validate = useCallback((values: SearchBarFormValues) => validate(values, limitDuration, setFieldWarning, validateQueryString, pluggableSearchBarControls, formatTime, handlerContext),
    [limitDuration, setFieldWarning, validateQueryString, pluggableSearchBarControls, formatTime, handlerContext]);

  return (
    <Formik<SearchBarFormValues> initialValues={_initialValues}
                                 enableReinitialize={enableReinitialize}
                                 onSubmit={_onSubmit}
                                 innerRef={formRef}
                                 validate={_validate}
                                 validateOnBlur={false}
                                 validateOnMount={validateOnMount}>
      {(...args) => (
        <Form>
          {_isFunction(children) ? children(...args) : children}
        </Form>
      )}
    </Formik>
  );
};

SearchBarForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  limitDuration: PropTypes.number.isRequired,
  validateOnMount: PropTypes.bool,
};

SearchBarForm.defaultProps = {
  validateOnMount: true,
  formRef: undefined,
};

export default SearchBarForm;
