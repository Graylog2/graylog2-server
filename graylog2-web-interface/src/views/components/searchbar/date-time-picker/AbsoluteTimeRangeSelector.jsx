// @flow strict
import * as React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'formik';
import styled, { css } from 'styled-components';
import type { StyledComponent } from 'styled-components';

import DateTime from 'logic/datetimes/DateTime';
import { Icon } from 'components/common';
import type { ThemeInterface } from 'theme';
import DateInputWithPicker from 'views/components/searchbar/DateInputWithPicker';

import TimerangeSelector from './TimerangeSelector';

type Props = {
  disabled: boolean,
};

const StyledTimerangeSelector: StyledComponent<{}, void, *> = styled(TimerangeSelector)`
  display: flex;
`;

const InputWrap: StyledComponent<{}, void, HTMLDivElement> = styled.div`
  width: 200px;
`;

const Separator: StyledComponent<{}, ThemeInterface, HTMLParagraphElement> = styled.p(({ theme }) => css`
  margin: 0;
  line-height: 34px;
  font-size: ${theme.fonts.size.large};
  padding-left: 15px;
  padding-right: 15px;
`);

const _isValidDateString = (dateString: string) => {
  if (dateString === undefined) {
    return undefined;
  }

  return DateTime.isValidDateString(dateString)
    ? undefined
    : 'Format must be: YYYY-MM-DD [HH:mm:ss[.SSS]]';
};

const AbsoluteTimeRangeSelector = ({ disabled }: Props) => {
  return (
    <StyledTimerangeSelector className="absolute">
      <Field name="tempTimeRange.from" validate={_isValidDateString}>
        {({ field: { value, onChange, onBlur, name }, meta: { error } }) => (
          <InputWrap>
            <DateInputWithPicker disabled={disabled}
                                 onChange={onChange}
                                 onBlur={onBlur}
                                 value={value}
                                 name={name}
                                 title="Search start date"
                                 error={error} />
          </InputWrap>
        )}
      </Field>

      <Separator className="text-center">
        <Icon name="long-arrow-alt-right" />
      </Separator>

      <Field name="tempTimeRange.to" validate={_isValidDateString}>
        {({ field: { value, onChange, onBlur, name }, meta: { error } }) => (
          <InputWrap>
            <DateInputWithPicker disabled={disabled}
                                 onChange={onChange}
                                 onBlur={onBlur}
                                 value={value}
                                 name={name}
                                 title="Search end date"
                                 error={error} />
          </InputWrap>
        )}
      </Field>
    </StyledTimerangeSelector>
  );
};

AbsoluteTimeRangeSelector.propTypes = {
  disabled: PropTypes.bool,
};

AbsoluteTimeRangeSelector.defaultProps = {
  disabled: false,
};

export default AbsoluteTimeRangeSelector;
