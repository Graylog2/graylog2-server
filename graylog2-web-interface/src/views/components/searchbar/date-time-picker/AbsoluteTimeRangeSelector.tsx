/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import type { StyledComponent } from 'styled-components';

import { TimeRange, AbsoluteTimeRange } from 'views/logic/queries/Query';
import { Icon } from 'components/common';

import AbsoluteRangeField from './AbsoluteRangeField';

type Props = {
  disabled: boolean,
  originalTimeRange: TimeRange,
  limitDuration: number,
  currentTimeRange: AbsoluteTimeRange,
};

const AbsoluteWrapper: StyledComponent<{}, void, HTMLDivElement> = styled.div`
  display: flex;
  align-items: stretch;
  justify-content: space-around;
`;

const RangeWrapper: StyledComponent<{}, void, HTMLDivElement> = styled.div`
  flex: 4;
  align-items: center;
  min-height: 290px;
  display: flex;
  flex-direction: column;
  
  .DayPicker-wrapper {
    padding-bottom: 0;
  }
`;

const IconWrap: StyledComponent<{}, void, HTMLDivElement> = styled.div`
  flex: 0.75;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const AbsoluteTimeRangeSelector = ({ disabled, limitDuration, originalTimeRange, currentTimeRange }: Props) => {
  return (
    <AbsoluteWrapper>
      <RangeWrapper>
        <AbsoluteRangeField from
                            originalTimeRange={originalTimeRange}
                            disabled={disabled}
                            limitDuration={limitDuration}
                            currentTimeRange={currentTimeRange} />
      </RangeWrapper>

      <IconWrap>
        <Icon name="arrow-right" />
      </IconWrap>

      <RangeWrapper>
        <AbsoluteRangeField from={false}
                            originalTimeRange={originalTimeRange}
                            disabled={disabled}
                            limitDuration={limitDuration}
                            currentTimeRange={currentTimeRange} />
      </RangeWrapper>
    </AbsoluteWrapper>
  );
};

AbsoluteTimeRangeSelector.propTypes = {
  disabled: PropTypes.bool,
  originalTimeRange: PropTypes.shape({ from: PropTypes.string, to: PropTypes.string }).isRequired,
  limitDuration: PropTypes.number,
  currentTimeRange: PropTypes.shape({ from: PropTypes.string, to: PropTypes.string }).isRequired,
};

AbsoluteTimeRangeSelector.defaultProps = {
  disabled: false,
  limitDuration: 0,
};

export default AbsoluteTimeRangeSelector;
