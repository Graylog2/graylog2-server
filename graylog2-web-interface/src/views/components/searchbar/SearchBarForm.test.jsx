// @flow strict
import * as React from 'react';
import { render } from 'wrappedTestingLibrary';

import SearchBarForm from './SearchBarForm';
import AbsoluteTimeRangeSelector from './date-time-picker/AbsoluteTimeRangeSelector';

describe('SearchBarForm', () => {
  describe('with AbsoluteTimeRangeSelector', () => {
    it('renders', () => {
      const initialValues = {
        timerange: { type: 'absolute', from: '2020-01-16 10:04:30.329', to: '2020-01-17 10:04:30.329' },
        queryString: '*',
        streams: [],
      };

      const { asFragment } = render(
        <SearchBarForm onSubmit={() => {}}
                       initialValues={initialValues}>
          <AbsoluteTimeRangeSelector />
        </SearchBarForm>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
