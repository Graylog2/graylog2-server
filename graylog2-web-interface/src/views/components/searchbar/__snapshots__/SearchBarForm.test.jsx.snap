// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBarForm with AbsoluteTimeRangeSelector renders 1`] = `
<DocumentFragment>
  .c0 {
  height: 100%;
}

.c3.btn-danger {
  background-color: #ad0707;
  border-color: #b23939;
  color: rgb(252,249,249);
  -webkit-transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
  transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
}

.c3.btn-danger:hover {
  background-color: #a00e0e;
  border-color: #a53636;
  color: #e9e6e6;
}

.c3.btn-danger.active {
  background-color: #b75151;
  border-color: #bc6363;
  color: rgb(252,249,249);
}

.c3.btn-danger.active:hover {
  background-color: #a94c4c;
  border-color: #ad5c5c;
  color: #e9e6e6;
}

.c3.btn-danger[disabled],
.c3.btn-danger.disabled {
  background-color: #c07272;
  border-color: #bc6363;
  color: rgb(252,250,250);
}

.c3.btn-danger[disabled]:hover,
.c3.btn-danger.disabled:hover {
  background-color: #c07272;
  border-color: #bc6363;
  color: rgb(252,250,250);
}

.c3.btn-default {
  background-color: #e6e6e6;
  border-color: #e0e0e0;
  color: rgb(73,73,73);
  -webkit-transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
  transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
}

.c3.btn-default:hover {
  background-color: #d4d4d4;
  border-color: #cfcfcf;
  color: #444444;
}

.c3.btn-default.active {
  background-color: #dadada;
  border-color: #d4d4d4;
  color: rgb(49,49,49);
}

.c3.btn-default.active:hover {
  background-color: #cacaca;
  border-color: #c4c4c4;
  color: #2f2f2f;
}

.c3.btn-default[disabled],
.c3.btn-default.disabled {
  background-color: #cecece;
  border-color: #d4d4d4;
  color: rgb(80,80,80);
}

.c3.btn-default[disabled]:hover,
.c3.btn-default.disabled:hover {
  background-color: #cecece;
  border-color: #d4d4d4;
  color: rgb(80,80,80);
}

.c3.btn-info {
  background-color: #0063be;
  border-color: #3970c2;
  color: rgb(249,250,252);
  -webkit-transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
  transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
}

.c3.btn-info:hover {
  background-color: #0c5cb0;
  border-color: #3668b3;
  color: #e6e7e9;
}

.c3.btn-info.active {
  background-color: #517cc5;
  border-color: #6386c9;
  color: rgb(249,250,252);
}

.c3.btn-info.active:hover {
  background-color: #4b73b6;
  border-color: #5c7dba;
  color: #e6e7e9;
}

.c3.btn-info[disabled],
.c3.btn-info.disabled {
  background-color: #7290cd;
  border-color: #6386c9;
  color: rgb(250,251,253);
}

.c3.btn-info[disabled]:hover,
.c3.btn-info.disabled:hover {
  background-color: #7290cd;
  border-color: #6386c9;
  color: rgb(250,251,253);
}

.c3.btn-link {
  background-color: rgba(255,255,255,0);
  border-color: rgba(255,255,255,0);
  color: #702785;
  -webkit-transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
  transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
}

.c3.btn-link:hover {
  background-color: transparent;
  border-color: transparent;
  color: #410057;
}

.c3.btn-link.active {
  background-color: rgba(255,255,255,0);
  border-color: rgba(255,255,255,0);
  color: #410057;
}

.c3.btn-link.active:hover {
  background-color: rgba(255,255,255,0);
  border-color: transparent;
  color: #410057;
}

.c3.btn-link[disabled],
.c3.btn-link.disabled {
  background-color: rgba(255,255,255,0);
  border-color: rgba(255,255,255,0);
  color: #702785;
}

.c3.btn-link[disabled]:hover,
.c3.btn-link.disabled:hover {
  background-color: rgba(255,255,255,0);
  border-color: rgba(255,255,255,0);
  color: #702785;
}

.c3.btn-primary {
  background-color: #702785;
  border-color: #7b458e;
  color: rgb(234,229,236);
  -webkit-transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
  transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
}

.c3.btn-primary:hover {
  background-color: #68267b;
  border-color: #724083;
  color: #d8d3da;
}

.c3.btn-primary.active {
  background-color: #855996;
  border-color: #8f699d;
  color: rgb(250,249,251);
}

.c3.btn-primary.active:hover {
  background-color: #7c538b;
  border-color: #846292;
  color: #e7e6e8;
}

.c3.btn-primary[disabled],
.c3.btn-primary.disabled {
  background-color: #9877a5;
  border-color: #8f699d;
  color: rgb(251,250,251);
}

.c3.btn-primary[disabled]:hover,
.c3.btn-primary.disabled:hover {
  background-color: #9877a5;
  border-color: #8f699d;
  color: rgb(251,250,251);
}

.c3.btn-success {
  background-color: #00ae42;
  border-color: #39b356;
  color: rgb(249,252,249);
  -webkit-transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
  transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
}

.c3.btn-success:hover {
  background-color: #0ca13e;
  border-color: #36a550;
  color: #e6e9e6;
}

.c3.btn-success.active {
  background-color: #51b866;
  border-color: #63bc74;
  color: rgb(249,252,250);
}

.c3.btn-success.active:hover {
  background-color: #4baa5f;
  border-color: #5cae6c;
  color: #e6e9e7;
}

.c3.btn-success[disabled],
.c3.btn-success.disabled {
  background-color: #72c180;
  border-color: #63bc74;
  color: rgb(250,252,250);
}

.c3.btn-success[disabled]:hover,
.c3.btn-success.disabled:hover {
  background-color: #72c180;
  border-color: #63bc74;
  color: rgb(250,252,250);
}

.c3.btn-warning {
  background-color: #ffd200;
  border-color: #f9cd07;
  color: rgb(57,47,0);
  -webkit-transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
  transition: background-color 150ms ease-in-out,border 150ms ease-in-out,color 150ms ease-in-out;
}

.c3.btn-warning:hover {
  background-color: #ebc20c;
  border-color: #e6bd0e;
  color: #362d0c;
}

.c3.btn-warning.active {
  background-color: #f2c70a;
  border-color: #ebc20c;
  color: rgb(54,45,2);
}

.c3.btn-warning.active:hover {
  background-color: #e0b80f;
  border-color: #d9b310;
  color: #332b0c;
}

.c3.btn-warning[disabled],
.c3.btn-warning.disabled {
  background-color: #e4bc0e;
  border-color: #ebc20c;
  color: rgb(88,73,5);
}

.c3.btn-warning[disabled]:hover,
.c3.btn-warning.disabled:hover {
  background-color: #e4bc0e;
  border-color: #ebc20c;
  color: rgb(88,73,5);
}

.c2.form-control:not([type="range"]) {
  color: #1f1f1f;
  background-color: #fff;
  border-color: #a6a6a6;
}

.c2.form-control:not([type="range"])::-webkit-input-placeholder {
  color: #9b9b9b;
}

.c2.form-control:not([type="range"])::-moz-placeholder {
  color: #9b9b9b;
}

.c2.form-control:not([type="range"]):-ms-input-placeholder {
  color: #9b9b9b;
}

.c2.form-control:not([type="range"])::placeholder {
  color: #9b9b9b;
}

.c2.form-control:not([type="range"]):focus {
  border-color: #7894ce;
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.075),0 0 8px rgba(120,148,206,0.4);
}

.c2.form-control:not([type="range"])[disabled],
.c2.form-control:not([type="range"])[readonly],
fieldset[disabled] .c2.form-control:not([type="range"]) {
  background-color: #dddddd;
  color: #9b9b9b;
}

.c2.form-control:not([type="range"]) ~ .form-control-feedback.glyphicon {
  display: none;
}

.c1 {
  width: 200px;
}

.c4 {
  margin: 0;
  line-height: 34px;
  font-size: 1.125rem;
  padding-left: 15px;
  padding-right: 15px;
}

<form
    action="#"
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        <div
          class="form-group"
        >
          <span>
            <span
              class="input-group"
            >
              <input
                autocomplete="off"
                class="c2 absolute form-control"
                id="date-input-tempTimeRange.from"
                label=""
                name="tempTimeRange.from"
                placeholder="YYYY-MM-DD HH:mm:ss"
                required=""
                type="text"
                value=""
              />
              <span
                class="input-group-btn"
              >
                <button
                  class="c3 btn btn-default"
                  title="Insert current date"
                  type="button"
                >
                  <svg
                    class="svg-inline--fa fa-magic"
                  />
                </button>
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
    <p
      class="c4 text-center"
    >
      <svg
        class="svg-inline--fa fa-long-arrow-alt-right"
      />
    </p>
    <div
      class="c1"
    >
      <div>
        <div
          class="form-group"
        >
          <span>
            <span
              class="input-group"
            >
              <input
                autocomplete="off"
                class="c2 absolute form-control"
                id="date-input-tempTimeRange.to"
                label=""
                name="tempTimeRange.to"
                placeholder="YYYY-MM-DD HH:mm:ss"
                required=""
                type="text"
                value=""
              />
              <span
                class="input-group-btn"
              >
                <button
                  class="c3 btn btn-default"
                  title="Insert current date"
                  type="button"
                >
                  <svg
                    class="svg-inline--fa fa-magic"
                  />
                </button>
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
